### PR TITLE
Implementation of allele_remap()

### LINKED ISSUE #2211: Create a function to map from one alleles list to another
I'm writing several functions to calculate metrics to assess imputation performance (see #2199 #2200 #2210). As pointed out in #2199 , when the alleles lists of two variants are not equal, the genotypes of the variants must be remapped so that the genotypes are comparable. Because this remapping needs to be done a lot to support #2199, #2200, and #2210, it may be good to have a separate function to do this remapping. @benjeffery and I chatted about adding such a function.

### PR DESCRIPTION
**Pull request by szhan**
Fixes #2211 

This is meant to be an internal function.

### CODE REVIEW

**Comment by szhan** [comment]
When committing, I got a message from ``flake8`` about having a bare ``except``. So, to work around this, I added ``Exception``. I was thinking to use ``IndexError`` instead, but that causes the function to terminate with an error. Is there a  better way to deal with a bare ``except`` here?

**Comment by benjeffery** [comment]
The exception thrown is `ValueError` so you should catch that:
```python
>>> l = [1,2]
>>> l.index(2)
1
>>> l.index(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: 3 is not in list
```

**Comment by szhan** [comment]
I'm trying to think of weird edge cases to test. But this function is quite simple, so I'm not sure if we need more tests. Should we add tests to check that ``alleles_to`` and ``alleles_from`` are Python ``list``?

File: python/tests/test_util.py [hunk 1]
```diff
@@ -540,3 +540,27 @@ def test_length_bad_value(self, length):
...
         with pytest.raises(TypeError, match="argument must be a string"):
             tskit.random_nucleotides(length)
+
+
+class TestAlleleRemap:
+    a = ["A", "C", "G", "T"]
+    b = ["G", "C"]
+    c = []
+
+    def test_a_bigger_than_b(self):
```
Comment (benjeffery) [inline python/tests/test_util.py:550]: Maybe rename these to `test_from_bigger_than_to` and `test_to_bigger_than_from` as currently `test_a_smaller_than_b` looks like an error as it isn't.

File: python/tests/test_util.py [hunk 2]
```diff
@@ -540,3 +540,27 @@ def test_length_bad_value(self, length):
     def test_length_bad_type(self, length):
         with pytest.raises(TypeError, match="argument must be a string"):
             tskit.random_nucleotides(length)
+
+
+class TestAlleleRemap:
+    a = ["A", "C", "G", "T"]
+    b = ["G", "C"]
+    c = []
```
Comment (benjeffery) [inline python/tests/test_util.py:548]: It would be good to add some more tests here you could use `@pytest.mark.parametrize` as in `test_reference_sequence:179`. Some ideas:
 - What happens if one/both the arguments are a single string?
 - Error states such as an argument being `None`, these tests help us understand what exceptions are thrown by a method and that they don't silently do something unwanted.

File: python/tskit/util.py [hunk 3]
```diff
@@ -592,3 +592,33 @@ def random_nucleotides(length: numbers.Number, *, seed: Union[int, None] = None)
...
+
+def allele_remap(alleles_from, alleles_to):
+    """
+    Returns an index map from the elements in one list (alleles_from)
+    to the elements of another list (alleles_to).
+
+    If some elements in alleles_from are not in alleles_to,
+    then indices outside of alleles_to are used.
+
+    :param list alleles_from: A list of alleles to map from.
```
Comment (benjeffery) [inline python/tskit/util.py:610]: The current type requirement for this function isn't `list` but `Sequence` as things like strings, tuples and bytes support the `.index` method.
I wonder if we should be supporting numpy arrays as a possible input to this function? I know `Variant.alleles` returns a `tuple`, so maybe we don't need to.

File: python/tskit/util.py [hunk 4]
```diff
@@ -592,3 +592,33 @@ def random_nucleotides(length: numbers.Number, *, seed: Union[int, None] = None)
...
     return a.tobytes().decode("ascii")
+
+
+#
+# Miscellaneous auxiliary methods.
+#
+
+
+def allele_remap(alleles_from, alleles_to):
+    """
```
Comment (benjeffery) [inline python/tskit/util.py:603]: I'm not 100% sure we want this method to be documented. Documenting a method declares it as something we think is useful externally and is a commitment to best-effort support indefinitely so it always has to be a considered action.

**Review by benjeffery — COMMENTED** [review]
Looking good! Some questions and comments.

**Comment by jeromekelleher** [comment]
Just a quick comment here on usability for this: we shouldn't be too strict about the input types, as we'd also like to be able to use this for comparing ``tsinfer.Variant`` objects with ``tskit.Variant``. I'm currently doing this and hitting the tedious allele shuffling problem.

File: python/tskit/util.py (hunk 4 above)
Comment (szhan) [reply python/tskit/util.py:603]: Is it good to keep two sentences about what it does? Or is too obvious enough?

```
"""
Returns an index map from the elements in one list (alleles_from) to the elements of another list (alleles_to).

If some elements in alleles_from are not in alleles_to, then indices outside of alleles_to are used.
"""
```

Comment (benjeffery) [reply python/tskit/util.py:603]: Yes, fine to keep that, but format it with `#` as comments so it doesn't get picked up by sphinx.

File: python/tests/test_util.py (hunk 2 above)
Comment (szhan) [reply python/tests/test_util.py]: By "one/both the arguments are a single string", do you mean if at least one of the inputs is not a list, just a string?

Comment (szhan) [reply python/tests/test_util.py]: Thanks, I get what you mean now!

File: python/tests/test_util.py [hunk 5]
```diff
@@ -540,3 +540,29 @@ def test_length_bad_value(self, length):
     def test_length_bad_type(self, length):
         with pytest.raises(TypeError, match="argument must be a string"):
             tskit.random_nucleotides(length)
+
+
+#
+# Tests for allele_remap
+#
+@pytest.mark.parametrize(
```
Comment (benjeffery) [inline python/tests/test_util.py:548]: Rather than have the same code repeated in three tests, combine all the `parametrize` lists, adding comments like "differs in length" and "both empty" before the line of those ones.

**Review by benjeffery — APPROVED** [review]
Sorry was waiting for @szhan to say this was ready.

One small niggle then we can merge.

File: python/tests/test_util.py (hunk 2 above)
Comment (szhan) [reply python/tests/test_util.py]: Maybe we can check that both the input lists are of the same type and are one of list, tuple or string.

Comment (benjeffery) [reply python/tests/test_util.py]: I don't think we need to be overly restrictive - they don't need to be the same type. Also I almost forgot - can we add some numpy examples to the tests?

Comment (szhan) [reply python/tests/test_util.py]: The function is using index(), which numpy arrays don't have. Maybe convert a numpy array into a Python list?

Comment (szhan) [reply python/tests/test_util.py]: Ah! You mentioned below!

Comment (benjeffery) [reply python/tests/test_util.py]: > The function is using `index()`, which numpy arrays don't have. Maybe convert a numpy array into a Python list?

Probably better to do it the other way around and convert to numpy array at the start, then we support all sequences that numpy does.

Comment (szhan) [reply python/tests/test_util.py]: So, the inputs can be a Numpy array or any Python Sequence-type object. And the output is a Numpy array (or a Python list)?

Comment (szhan) [reply python/tests/test_util.py]: Also, we are allowing for duplicate elements in either or both of the inputs?

Comment (szhan) [reply python/tests/test_util.py]: I've updated the function. Now, the inputs can be a numpy array or any Python Sequence-type object, and the output is a numpy array.

Comment (szhan) [reply python/tests/test_util.py]: When the input is a string, then it is treated as a list with one element, which is the string itself.

File: python/tests/test_util.py [hunk 6]
```diff
@@ -548,15 +548,19 @@ def test_length_bad_type(self, length):
...
+        # Case 2: alleles_to is shorter than alleles_from.
+        (["G", "C"], ["A", "C", "G", "T"], np.array([2, 1], dtype="uint32")),
+        # Case 3: alleles_to is empty.
+        (["A", "C", "G", "T"], [], np.array([0, 1, 2, 3], dtype="uint32")),
+        # Case 4: alleles_from is empty.
+        ([], ["A", "C", "G", "T"], np.array([], dtype="uint32")),
+        # Case 5: Both lists are empty.
+        ([], [], np.array([], dtype="uint32")),
+        # Case 6: Both lists are tuples.
+        (("G", "C"), ("A", "C", "G", "T"), np.array([2, 1], dtype="uint32")),
```
Comment (jeromekelleher) [inline python/tests/test_util.py:562]: What about multi-character alleles? We need to be able to handle arbitrary strings as alleles. Like ``["AAA", "AA", "A", ""]``, e.g.

Comment (szhan) [reply python/tests/test_util.py:562]: Added a test for that.

Comment (benjeffery) [reply python/tests/test_util.py:562]: Oh and one more - as we support arbitrary unicode in these fields it would be good to have a test with the classic poo emoji as one of the alleles.

**Comment by benjeffery** [comment]
@szhan Is this ready for another look? I can also clean up the commits if you like? Seems to have a merge in at the moment.

**Comment by szhan** [comment]
Cancelling and closing this PR. Merging these commits into my own local fork caused too many weird conflicts, so it is far easier to just redo it.
